### PR TITLE
275 support hreflang

### PR DESCRIPTION
**Pull request by UlrichKu**
| Q             | A
| ------------- | ---
| Branch        | 6.2.x
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed issues  | chameleon-system/chameleon-system#275
| License       | MIT

### CODE REVIEW

File: src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php [hunk 1]
```diff
@@ -647,33 +650,101 @@ protected function GetETrackerId($oActivePage)
...
+        if (null === $activeLanguage) {
+            return [];
+        }
+
+        $activeLanguages = $activePortal->GetActiveLanguages();
+        if ($activeLanguages->Length() < 2) {
+            return [];
+        }
+
+        $activePage = $this->getActivePageService()->getActivePage();
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php]: ok

File: src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php [hunk 2]
```diff
@@ -647,33 +650,101 @@ protected function GetETrackerId($oActivePage)
...
+        if (true === $this->isNotFoundPage($activePage, $activePortal)) {
+            return [];
+        }
+
+        $alternatives = [];
+        while (false !== ($alternativeLanguage = $activeLanguages->Next())) {
+            if ($alternativeLanguage->id !== $activeLanguage->id) {
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php]: ok

File: src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php [hunk 3]
```diff
@@ -647,33 +650,101 @@ protected function GetETrackerId($oActivePage)
...
+                } catch (Exception $exception) {
+                    $id = '';
+                    $name = '';
+                    if (null !== $activePage) {
+                        $id = $activePage->id;
+                        $name = $activePage->GetName();
+                    }
+
+                    $this->getLogger()->error(
+                        sprintf('Cannot generate alternative language urls for page %s/%s with language %s.', $id, $name, $iso),
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php]: ok

File: src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php [hunk 4]
```diff
@@ -647,33 +650,101 @@ protected function GetETrackerId($oActivePage)
...
+    }
+
+    private function getLanguageService(): LanguageServiceInterface
+    {
+        return ServiceLocator::get('chameleon_system_core.language_service');
+    }
+
+    private function getLogger(): LoggerInterface
+    {
+        // TODO already use this with 6.2?
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/private/library/classes/WebModules/MTPageMetaCore/MTPageMetaCoreEndPoint.class.php]: removed

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1550235650.inc.php [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<h1>Build #1550235650</h1>
+<h2>Date: 2019-02-15</h2>
+<div class="changelog">
+    - #275: Add info message for hreflang
+</div>
+<?php
+
+TCMSLogChange::addInfoMessage('hreflang: MTPageMeta provides now the array "language-alternatives" with URLs to the current page in other languages.
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1550235650.inc.php]: How about something like this:

"MTPageMeta now provides an additional variable "language-alternatives" that allows to add URLs to alternative languages of the same page using hreflang meta tags. See the view MTPageMeta/standard.view.php in chameleon-shop-theme-bundle for an example."

And I'd add the message on INFO level; it's an optional improvement, so not really a TODO.

Comment (UlrichKu) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1550235650.inc.php]: ok, ok

**Comment by umulmrum** [comment]
Thanks @UlrichKu

**Pull request merged**